pkg/bgp/speaker/events.go

```diff
@@ -49,18 +69,57 @@ func (s *Speaker) run() {
 // do performs the appropriate action depending on the event type. For example,
 // if it is a service event (svcEvent), then it will call into MetalLB's
 // SetService() to perform BGP announcements.
-func (s *Speaker) do(key interface{}) types.SyncState {
+func (s *MetalLBSpeaker) do(key interface{}) types.SyncState {
 	defer s.queue.Done(key)
 
 	switch k := key.(type) {
 	case svcEvent:
-		return s.SetService(s.logger, k.id.String(), k.svc, k.eps)
+		return s.speaker.SetService(k.id.String(), k.svc, k.eps)
 	case epEvent:
-		return s.SetService(s.logger, k.id.String(), k.svc, k.eps)
+		return s.speaker.SetService(k.id.String(), k.svc, k.eps)
 	case nodeEvent:
-		return s.SetNodeLabels(s.logger, *k)
+		return s.handleNodeEvent(k)
 	default:
 		log.Debugf("Encountered an unknown key type %T in BGP speaker", k)
 		return types.SyncStateSuccess
 	}
 }
+
+func (s *MetalLBSpeaker) handleNodeEvent(k nodeEvent) types.SyncState {
+	var (
+		ret    types.SyncState
+		failed bool
+	)
+
+	if k.withDraw {
+		// this is a best effort method call, so we don't
+		// care about errors. If the node is shutting down
+		// all routes will be closed once the speaker's TCP conn
+		// is closed anyway per the rfc.
+		// see: https://datatracker.ietf.org/doc/html/rfc4271#section-6
+		s.withDraw()
+		atomic.AddInt32(&s.shutdown, 1)
+		return types.SyncStateSuccess
+	}
+
+	if s.announceLBIP {
+		if r := s.speaker.SetNodeLabels(*k.labels); r != types.SyncStateSuccess {
+			failed = true
+			ret = r
+		}
+	}
+	if s.announcePodCIDR {
+		if err := s.announcePodCIDRs(*k.podCIDRs); err != nil {
+			if !failed {
+				failed = true
+				ret = types.SyncStateError
+			}
+			log.WithError(err).WithField("CIDRs", k.podCIDRs).Error("Failed to announce pod CIDRs")
+		}
+	}
+
+	if failed {
```

Comment:
unimportant nit: could simplify this if `ret = types.SyncStateSuccess` at beginning and doing just `return ret` here

Reply:
@christarazi, @joamaki comment makes me think here. 

This function `handleNodeEvent` has two side effects:
- set node labels 
- announce pod cidr

Because this forms a permutation of results, we wind up in the situation where we may return "SyncStateFailed" even when one of those side effects have occurred. This feels wrong to me. 

I think we may want to realize the `handleNodeEvent` as a atomic operation (loosely speaking) where if any of the configured side effects fails, we bail immediately. What do you think Chris and Jussi?

Reply:
looking at this a bit more, I'm noticing it is this way because we reuse MetalLB's type `SyncState` as MetalLB doesn't return error codes proper, but seems to summarize things with that enum. That's a little unfortunate because it lacks detail for us. 

That being said, its probably fine for now, the only real rub with this is that if say, announcing the pod cidr fails in a loop, but setting the node labels is successful, we will continually set node labels everytime we retry the pod cidr announcement.

Reply:
Yeah it's not as clean as it should be, but the risk for problems by setting the node labels is very low.

Reply:
Yeah, I'm fine with this.